External Initiator for Klaytn

winnie-byun opened the pull request:
External Initiator for Klaytn.

`subscribe` API in Klaytn works similar with eth's. It is different from eth that klaytn uses 'klay' instead of 'eth' in method.
If you are subscribing something other than Log, it could have different struct from eth's. (e.g. Structs such as Headers are different.)

I've tested Klaytn EI with chainlink node and  attaching to Klaytn testnet (Baobab)
: https://github.com/winnie-byun/demo-klaytn-chainlink

aalu1418 commented:
@winnie-byun Hello! I've been assigned to review the Klatyn EI implementation. After a first look, here's some feedback so we can get your PR merged!

* It seems like you based the `klatyn.go` file from the `eth.go` file which is a good start. There are some functions in creating the filter query and parsing the event logs that would be needed to make it work with our on-chain smart contracts (example from the BSC initiator).
  * Creating the filter query: https://github.com/smartcontractkit/external-initiator/blob/master/blockchain/binance-smart-chain.go#L27
  * Parsing: https://github.com/smartcontractkit/external-initiator/blob/master/blockchain/binance-smart-chain.go#L133
* Make sure all the integration tests run successfully. They can be run with instructions in the root readme
* There also seems to be some linting errors, these can usually be solved with `go fmt`
* Also, please rebase so the latest changes are included

Thanks!

winnie-byun commented:
@aalu1418 I changed the code like the BSC initiator and rebased.
The code is already `go fmt`ed.
I tested that with a chainlink core and it runs with no error. I will update after I tested with klaytn chainlink adapter.
Please take a look.

aalu1418 commented on integration/mock-client/blockchain/klaytn_test.go:
```diff
@@ -0,0 +1,192 @@
+			}
+		})
+	}
+}
```
Please add testing functions for `klaytnLogRequestToResponse` and `handleKlaytnGetLogs` 😄

aalu1418 commented on blockchain/klaytn.go:
```diff
@@ -0,0 +1,204 @@
+	}
+
+	return events, true
+}
```
Seems like you're missing the `blockchain/klaytn_test.go` file for testing the various functions?

aalu1418 reviewed (changes requested):
Looking pretty good!
Most of the feedback is around some missing pieces that I noticed.

aalu1418 commented:
Requested changes will be addressed in a separate PR
